db/data_listeners.cc

```diff
@@ -84,7 +84,10 @@ flat_mutation_reader toppartitions_data_listener::on_read(const schema_ptr& s, c
         return std::move(rd);
     }
     dblog.trace("toppartitions_data_listener::on_read: {}.{}", s->ks_name(), s->cf_name());
-    return make_filtering_reader(std::move(rd), [this, &range, &slice, s = std::move(s)] (const dht::decorated_key& dk) {
+    // Cannot use with_gate() because it returns a future, and we need to return a value
+    _gate.enter();
+    auto leave = defer([this] { _gate.leave(); });
+    return make_filtering_reader(std::move(rd), [this, &range, &slice, s = std::move(s), leave = std::move(leave)] (const dht::decorated_key& dk) {
```

Comment:
There is no guarantee, but it won't be blocked indefinitely. The reader will either be exhausted or evicted due to idling for too long.

I see that stop() is called after results are gathered, but it can delay them, so we should do something. I guess by weak_ref you meant that the reader will hold a weak reference to _top_k_read, so we can detach top_k before the readers are all done?

Reply:
Do we actually currently evict due to idling?

Regarding weak_ref, yes, that's what I meant.

Reply:
@denesb do we?

Reply:
We have time based cache eviction, so yes.

Reply:
How long do we wait?

Reply:
10s if memory serves correctly.

Reply:
But reads can reset the timer, so if a paging client reads slowly, it can be much longer.

Reply:
Yes, perhaps better to use a weak ref.